autoload/sj/r.vim

```diff
@@ -0,0 +1,190 @@
+  let lines = split(text, "\n")
+
+  return lines
+endfunction
+
+" function! s:IsValidSelection(motion)
+"
+" Test whether a visual selection contains more than a single character after
+" performing the given normal-mode motion string
+"
+function! s:IsValidSelection(motion)
+  call sj#PushCursor()
+  execute "silent normal! " . a:motion . "\<esc>"
+  execute "silent normal! \<esc>"
+  let is_valid = getpos("'<") != getpos("'>")
+  call sj#PopCursor()
+  return is_valid
+endfunction
+
+" function! s:ReplaceMotionPreserveCursor(motion, lines [, inserts [, mask]]) {{{2
+"
+" Replace the normal mode "motion" with a list of "lines", separated by line
+" breaks, and optionally "inserts" characters, while making a best attempt at
+" preserving the cursor's location within the text block if it's replaced with
+" similar text. Optionally, a "mask" can be provided which is a boolean list
+" indication which text in "lines" originate as part of the original text of the
+" visual selection.
+"
+function! s:ReplaceMotionPreserveCursor(motion, rep, ...)
+  " default to interpretting all lines of text as originally from text to replace
+  let rep = a:rep
+  let mask = a:0 > 2 ? a:2 : repeat([1], len(a:rep))
+
+  " do motion and get bounds & text
+  call sj#PushCursor()
+  execute "silent normal! " . a:motion . "\<esc>"
+  execute "silent normal! \<esc>"
+  call sj#PopCursor()
+  let ini = map(s:GetTextRange("'<", "."), {k, v -> s:Triml(v)[0]})
+
+  " do replacement
+  let body = join(a:rep, "\n")
+  call sj#ReplaceMotion(a:motion, body)
+
+  " go back to start of selection
+  silent normal! `<
+
+  " try to reconcile initial selection against replacement lines
+  let [cursory, cursorx] = [0, 0]
+  while len(ini) && len(rep)
+    " rep[0] (next replacement line) should be present in initial selection
+    if mask[0]
+      let i = stridx(ini[0], rep[0])
+      let j = stridx(rep[0], ini[0])
+      if i >= 0
+        " if an entire line of the replacement text found in initial then we'll
+        " need our cursor to move to the next line if more lines are insered
+        let [ini[0], ws] = s:Triml(ini[0][i+len(rep[0]):])
+        let cursorx += i + len(rep[0])
+        let ini = len(ini[0]) ? ini : ini[1:]
+        let rep = rep[1:]
+        if len(ini)
+          let cursory += 1
+          let cursorx = 0
+        endif
+      elseif j >= 0
+        " if an entire line of the initial is found in the replacement then
+        " we'll need our cursor to move rightward through length of the initial
+        let [rep[0], ws] = s:Triml(rep[0][j+len(ini[0]):])
+        let cursorx += j + len(ini[0])
+        let ini = ini[1:]
+        let cursorx += (len(ini) && len(ini[0]) ? ws : 0)
+      else
+        let ini = []
+      endif
+      " continue to next rep (replacement line)
+    else
+      let rep = rep[1:]
+    endif
+  endwhile
+
+  call s:MoveCursor(cursory, max([cursorx-1, 0]))
+  call sj#PushCursor()
+endfunction
```

Comment:
It's an interesting approach, and it does seem to work reliably for these cases. Still, I've left it as a private/script-local function, because it's not used anywhere else, and I'd really like to have multiple use cases to decide what a sensible interface for it could be.

For instance, neither the `mask` and `inserts` are being used right now. Is an array of booleans a useful interface for `mask`? How are `inserts` going to be figured out, since they're not used at all right now? I would avoid these extra parameters until they're actually used in practice.

But as a private function, this is fine -- if and when I end up using it for something else (or get a PR using it), it might be a good point to move it to the shared `sj.vim` file.

Reply:
Ah, apologies for letting those slip through. I had used them at one point in the implementation and had forgotten to go back through and clean up the interfaces.

The intention was that `mask` could be used to indicate when a part of the replacement draws its origin from the original text. For some splits/joins the content of the code changes, so it's necessary to mask portions out from being used to try to align the cursor. `inserts` was used in conjunction with this to specify how many characters need to be inserted between lines to properly align the cursor. An example might be:

```vim
let x = len(a) ? "true" : "false"

" after split, becomes:
if len(a)
  let x = "true"
else 
  let x = "false"
endif
```

I never settled on something that I was particularly happy with to accommodate more comprehensive rearrangements and ended up scrapping the attempts. 

I'll clear these vestigial bits out and leave it for another exercise to try to handle non-whitespace or non-sequential rearrangments.